File: algorithms/code/leetcode/lc075_sort_colors/lc075_sort_colors.py
# https://leetcode.com/problems/sort-colors/

import unittest


class Solution:
    def sortColors(self, nums: list[int]) -> None:
        counters = [0, 0, 0]
        for value in nums:
            counters[value] += 1
        i = 0
        j = 0
        while i < len(counters):
            k = 0
            while k < counters[i]:
                nums[j] = i
                k += 1
                j += 1
            i += 1


class TestCode(unittest.TestCase):
    def test_2(self) -> None:
        nums = [2]
        expected = [2]
        Solution().sortColors(nums)
        self.assertListEqual(expected, nums)

    def test_1_0(self) -> None:
        nums = [1, 0]
        expected = [0, 1]
        Solution().sortColors(nums)
        self.assertListEqual(expected, nums)

File: algorithms/code/leetcode/lc075_sort_colors/lc075_sort_colors.rs
// https://leetcode.com/problems/sort-colors/
// #medium

pub struct Solution;

impl Solution {
    #[allow(clippy::ptr_arg)]
    pub fn sort_colors(nums: &mut Vec<i32>) {
        let mut counters = [0; 3];
        for value in nums.iter() {
            counters[*value as usize] += 1;
        }
        let mut j = 0;
        for (i, counter_i) in counters.iter().enumerate() {
            for _ in 0..*counter_i {
                nums[j] = i as i32;
                j += 1;
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_2() {
        let mut nums = vec![2];
        let expected = vec![2];
        Solution::sort_colors(&mut nums);
        assert_eq!(nums, expected);
    }

    #[test]
    fn test_10() {
        let mut nums = vec![1, 0];
        let expected = vec![0, 1];
        Solution::sort_colors(&mut nums);
        assert_eq!(nums, expected);
    }
}
